### PR TITLE
feat(rules): new resource type: synonyms and a massive refactoring

### LINKED ISSUE #521: new resource type: synonyms
As discussed in https://gapminder.slack.com/archives/C0FTBA1P0/p1526574683000058, we need to accept a new resource type `synonym`. 

A few rules I think should be implemented:

- primary key should be a list of 2 columns, one of them must be "synonym"
- there should be no duplicated keys

### PR DESCRIPTION
**Pull request by buchslava**
+ full code refactoring

Closes #521

### CODE REVIEW

File: src/data/data-package.ts [hunk 1]
```diff
@@ -467,100 +448,38 @@ export class DataPackage {
...
-              }).catch(err => {
-                throw err;
-              });
-            });
-        } catch (contentErr) {
-          this.errors.push({
-            source: contentErr,
-            reason: `${DATA_PACKAGE_FILE} parsing`
+      this.dataPackageContent = JSON.parse(content);
+      this.getTranslationFileDescriptors(
```
Comment (korel-san) [inline src/data/data-package.ts]: @buchslava async function move from try/catch

File: src/data/data-package.ts [hunk 2]
```diff
@@ -570,36 +489,64 @@ export class DataPackage {
...
 
+    this.isNewDataPackage = false;
+
     fileExists(filePath, (err, isExists) => {
       if (err || !isExists || ignoreExistingDataPackage) {
-        this.build(dataPackage => onDataPackageReady(dataPackage));
-        return;
-      }
+        this.isNewDataPackage = true;
+        this.buildDataPackageResources(dataPackage => onDataPackageReady(dataPackage));
```
Comment (korel-san) [inline src/data/data-package.ts]: @buchslava tracking build error

File: src/data/file-descriptor.ts [hunk 3]
```diff
@@ -89,33 +61,27 @@ export class FileDescriptor {
...
-        throw err;
+    this.checkFile((fileStateDescriptor) => {
+      if (!fileStateDescriptor.ok) {
+        this.issues.push(fileStateDescriptor);
+        return onFileDescriptorChecked([fileStateDescriptor]);
       }
 
       getFileLine(this.fullPath, 1, (lineErr, line) => {
-        const fileAttributesContainer: any = results.find((result: any) => result.ok);
+        const fileAttributesContainer = fileStateDescriptor;
```
Comment (korel-san) [inline src/data/file-descriptor.ts]: @buchslava log error

File: src/utils/file.ts [hunk 4]
```diff
@@ -148,14 +148,20 @@ export function readFile(filePath, onFileRead) {
 }
 
 export function walkFile(filePath, onLineRead, onFileRead) {
-  let line = 0;
+  fileExists(filePath, (err, exists) => {
+    if (err || !exists) {
+      return onFileRead();
```
Comment (korel-san) [inline src/utils/file.ts]: @buchslava log error

File: src/utils/file.ts [hunk 5]
```diff
@@ -148,14 +148,20 @@ export function readFile(filePath, onFileRead) {
...
-    .on('data', ddfRecord => onLineRead(ddfRecord, line++))
-    .on('end', () => {
-      onFileRead();
-    });
+    let line = 0;
+
+    csv
+      .fromPath(filePath, {headers: true})
+      .on('data', ddfRecord => onLineRead(ddfRecord, line++))
+      .on('end', () => {
```
Comment (korel-san) [inline src/utils/file.ts]: @buchslava on error

File: src/ddf-rules/general-rules/duplicated-synonym-key.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import * as path from 'path';
+import { head, keys } from 'lodash';
+import { DUPLICATED_SYNONYM_KEY } from '../registry';
+import { DdfDataSet } from '../../ddf-definitions/ddf-data-set';
+import { Issue } from '../issue';
+
+const getPrimaryKeyByFile = (ddfDataSet: DdfDataSet, file: string): string[] => <string[]>head(
+  ddfDataSet.getDataPackageResources()
+    .filter(resource => resource.path === file)
```
[the remark this answers is not in the document]
Comment (buchslava) [reply src/ddf-rules/general-rules/duplicated-synonym-key.ts]: no, because in this case, primaryKey is an array of strings. so primaryKey, not primaryKeys

**Pull request merged**